## Task 1
Data Cleaning Challenge on Sales Data.


In [45]:

import pandas as pd

# Read files
sales_2017 = pd.read_csv("Sales_Transactions_2017.csv")
sales_2018 = pd.read_csv("Sales_Transactions_2018.csv")
sales_2019 = pd.read_csv("Sales_Transactions_2019.csv")

# Combine all files
sales_all = pd.concat(
    [sales_2017, sales_2018, sales_2019],
    ignore_index=True
)


In [46]:
type(sales_all)

pandas.core.frame.DataFrame

In [47]:
sales_all.describe()

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
count,98615,98649,111166,98615,98649,98648,98648,5597,27560
unique,836,10044,1994,867,546,1109,2548,371,7149
top,,,TP13,,100,,,,
freq,3053,3053,13056,3053,12528,3051,3053,3053,3053


In [48]:
sales_all.head()

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
0,1/4/2017,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2,"1,690.00","3,380.00",NaN,"13,100.00"
1,1/4/2017,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
2,1/4/2017,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23,"11,500.00",NaN,"30,990.00"
3,1/4/2017,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
4,1/4/2017,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5,"1,690.00","8,450.00",NaN,NaN


In [49]:
sales_all.shape

(111206, 9)

### Removing duplicates - After removinig the duplicates the records count decreased from 111206 to 96624.

In [50]:
sales_all.drop_duplicates()

,Date,Voucher,Party,Product,Qty,Rate,Gross,Disc,Voucher Amount
0,1/4/2017,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2,"1,690.00","3,380.00",NaN,"13,100.00"
1,1/4/2017,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
2,1/4/2017,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23,"11,500.00",NaN,"30,990.00"
3,1/4/2017,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6,"1,620.00","9,720.00",NaN,NaN
4,1/4/2017,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5,"1,690.00","8,450.00",NaN,NaN
...,...,...,...,...,...,...,...,...,...
111199,10/10/2019,Sal:4935,K.SRIHARI,13*16 Bk(100)KRISHN,320,16,"5,120.00",NaN,NaN
111200,10/10/2019,Sal:4935,K.SRIHARI,10*12 RK,800,8.5,"6,800.00",NaN,NaN
111201,10/10/2019,Sal:4935,K.SRIHARI,13*16 WHITE RK,400,16,"6,400.00",NaN,NaN
111204,NaN,Total,NaN,NaN,"99,284.90","175,381.65","2,203,649.50","20,680.00","2,189,014.50"


In [51]:
# Clean column names

sales_all.columns = sales_all.columns.str.strip()


In [60]:

# View columns
sales_all.columns


Index(['Date', 'Voucher', 'Party', 'Product', 'Qty', 'Rate'], dtype='object')

In [53]:

# Keep only required columns
sales_all = sales_all[
    ['Date', 'Voucher', 'Party', 'Product', 'Qty', 'Rate']
]


In [63]:
sales_all.columns


Index(['Date', 'Voucher', 'Party', 'Product', 'Qty', 'Rate'], dtype='object')

In [55]:
sales_all.head()

,Date,Voucher,Party,Product,Qty,Rate
0,1/4/2017,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2,"1,690.00"
1,1/4/2017,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6,"1,620.00"
2,1/4/2017,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23
3,1/4/2017,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6,"1,620.00"
4,1/4/2017,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5,"1,690.00"


In [58]:

# Remove rows missing important fields
sales_all = sales_all.dropna(
    subset=['Date', 'Voucher', 'Party', 'Product', 'Qty', 'Rate']
)


In [59]:
sales_all

,Date,Voucher,Party,Product,Qty,Rate
0,1/4/2017,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2,"1,690.00"
1,1/4/2017,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6,"1,620.00"
2,1/4/2017,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23
3,1/4/2017,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6,"1,620.00"
4,1/4/2017,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5,"1,690.00"
...,...,...,...,...,...,...
111197,10/10/2019,Sal:4935,K.SRIHARI,16*20(100-W),140,26
111198,10/10/2019,Sal:4935,K.SRIHARI,10*12 KRISHNA-BK(10,600,8.4
111199,10/10/2019,Sal:4935,K.SRIHARI,13*16 Bk(100)KRISHN,320,16
111200,10/10/2019,Sal:4935,K.SRIHARI,10*12 RK,800,8.5


In [65]:
sales_all.dtypes

Date       object
Voucher    object
Party      object
Product    object
Qty        object
Rate       object
dtype: object

### Converting the Data type of Date to DateTime

In [66]:
sales_all['Date'] = pd.to_datetime(
    sales_all['Date'],
    dayfirst=True,
    errors='coerce'
)

In [67]:
print(sales_all['Date'].dtype)

datetime64[ns]


In [68]:
sales_all.dtypes

Date       datetime64[ns]
Voucher            object
Party              object
Product            object
Qty                object
Rate               object
dtype: object

In [70]:
sales_all.head()

,Date,Voucher,Party,Product,Qty,Rate
0,2017-04-01,Sal:1,SOLANKI PLASTICS,DONA-VAI-9100,2,"1,690.00"
1,2017-04-01,Sal:1,SOLANKI PLASTICS,LITE FOAM(1200),6,"1,620.00"
2,2017-04-01,Sal:2,SARNESWARA TRADERS,VISHNU CHOTA WINE,500,23
3,2017-04-01,Sal:2,SARNESWARA TRADERS,LITE FOAM(1200),6,"1,620.00"
4,2017-04-01,Sal:2,SARNESWARA TRADERS,DONA-VAI-9100,5,"1,690.00"


In [71]:
sales_all.isna().sum()

Date       3053
Voucher       0
Party         0
Product       0
Qty           0
Rate          0
dtype: int64

## Task 2
Merge Function 1 Example


### merge() is used to combine two DataFrames based on one or more common columns, just like SQL JOINs.

In [2]:
import pandas as pd

# Creating Sample DataFrames

user_details=pd.DataFrame({"ID":[1,2,3],
                          "Name":["Sujin","Divo","Ani"]})

city_details = pd.DataFrame({"ID":[1,2,4],
                            "City":["Chennai","Mumbai","Goa"]})

user_details

,ID,Name
0,1,Sujin
1,2,Divo
2,3,Ani


In [3]:
city_details

,ID,City
0,1,Chennai
1,2,Mumbai
2,4,Goa


### 1. Inner Merge (Default)
Returns only matching rows.

In [7]:
result = pd.merge(user_details,city_details,on="ID")
result

,ID,Name,City
0,1,Sujin,Chennai
1,2,Divo,Mumbai


### 2. Left Merge
Keeps all rows from the left DataFrame.

In [5]:
result=pd.merge(city_details,user_details,on="ID",how="left")

result

,ID,City,Name
0,1,Chennai,Sujin
1,2,Mumbai,Divo
2,4,Goa,NaN


### Right Merge
Keeps all rows from the right DataFrame.

In [83]:
result = pd.merge(user_details,city_details,on="ID",how="right")
result

,ID,Name,City
0,1,Sujin,Chennai
1,2,Divo,Mumbai
2,4,NaN,Goa


### Outer Merge
Keeps all rows from both DataFrames.

In [85]:
result = pd.merge(user_details,city_details,on="ID",how="outer")
result

,ID,Name,City
0,1,Sujin,Chennai
1,2,Divo,Mumbai
2,3,Ani,NaN
3,4,NaN,Goa


## Task 3
List Comprehension 2 examples


In [87]:
# Question 1: Square Even Numbers
#Given a list of numbers, create a new list containing the squares of only the even numbers using list comprehension.

output = [x**2 for x in range(1,11) if x%2==0]
output

[4, 16, 36, 64, 100]

In [92]:
# Question 2: Extract Long Words
#Given a list of words, create a new list containing only the words that have more than 5 characters using list comprehension.

words = ["apple", "banana", "kiwi", "strawberry", "mango"]


output =[word for word in words if len(word) > 5]

output

['banana', 'strawberry']

## Task 4
Lambda Function 2 examples with Map, Filter, Reduce


## map() Examples
map() applies a function to every item in an iterable.

In [93]:

# Example 1: Square Numbers

numbers = [1,2,3,4,5]

result= list(map(lambda x:x**2,numbers))
result

[1, 4, 9, 16, 25]

In [94]:
# Example 2: Convert Names to Uppercase

names = ["sujin","divo","ani"]
result= list(map(lambda x:x.upper(),names))

result

['SUJIN', 'DIVO', 'ANI']

### 2. filter() Examples
filter() selects elements that satisfy a condition.

In [98]:
# Example 1: Filter Even Numbers

numbers=[1,2,3,4,5,6,7,8,9,10]

result = list(filter(lambda x: x%2 ==0,numbers))
result

[2, 4, 6, 8, 10]

In [100]:
# Example 2: Filter Words Longer Than 5 Characters

words = ["apple", "banana", "kiwi", "strawberry", "mango"]

result = list(filter(lambda x:len(x) > 5,words))

result

['banana', 'strawberry']

### 3. reduce() Examples
reduce() applies a function cumulatively to reduce the iterable to a single value.

In [104]:
# Example 1: Sum of Numbers

from functools import reduce

numbers = [1,2,3,4,5,6,7,8,9]

output = reduce(lambda x,y:x+y,numbers)

output

45

In [105]:
# Example 2: Find Maximum Number

from functools import reduce

numbers = [1,2,3,4,5,6,7,8,9]

output = reduce(lambda x,y: x if x>y else y,numbers)

output


9